AP_NavEKF3: Accumulate error for cores

harshitsankhla opened the pull request:
To improve core selection logic, accumulate error for each EKF lane then select the one with lowest error as new primary.

We start by simply adding up NavEKF3_core::errorScore() for each lane. We would like to explore a better metric that takes into account - 
1. State Difference
2. Innovations
3. Variances

Finally, the error should be some weighted combination of factors can quantify a core's error.

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3_Outputs.cpp:
```diff
@@ -51,6 +51,16 @@ float NavEKF3_core::errorScore() const
 
+void NavEKF3_core::updateCoreError()
+{
+    runningError += NavEKF3_core::errorScore();
```
this looks like runningError will always increase. I think we need a number that is on the same scale throughout the flight, so we can have meaningful thresholds.
If you leave the log up in AP_NavEKF3.cpp then I think you could accumulate the difference between the error score of each lane and the error score of the primary. That number would always be zero for the primary, and would grow as +ve or -ve number for lanes that are worse or better than the primary.

harshitsankhla replied:
Yes, the unbounded error is an issue, I was thinking about some normalisation to constrain it in a range but could not figure out logically. I'll modify this to accumulating differences with respect to the primary. Also when we switch lanes, should I reset all the errors by making the new primary's error 0 or continue with the existing, which also is fine as error would still accumulate in the respective lanes, just there would be an initial offset which I don't know if it would create any problems.

tridge replied:
I think reset all to zero on switch

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -815,23 +815,29 @@ void NavEKF3::UpdateFilter(void)
+    }
+
     float primaryErrorScore = core[primary].errorScore();
     if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection) {
```
this if statement means we only consider lane changes if the primary is unhealthy or it's score is high. I think we need to consider switching if the accumulated difference between the primary and an alternative lane goes over a threshold

harshitsankhla replied:
I think that logic is this one just below it - 
`if ((primaryErrorScore > 1.0f || !core[primary].healthy()) && runCoreSelection)`
what the new if condition stands for is to start accumulating errors only after core selection logic is available (the if block just above it does that). The reason for waiting till core selection is available is to not accumulate errors that happen due to the initialisation process. Please correct me if I'm wrong. 
I think we should add a separate check that ensures even when the current primary is healthy or is returning an NavEKF_core::errorScore() < 1.0f,  if difference between accumulated error between primary and another lane is high above a threshold, we should switch.

tridge replied:
sounds good, thanks

harshitsankhla commented:
Status as on June 16, 2020 -

1. Introduced error accumulation for all active cores, relative to the primary core at the instance. Reset core errors to zero when we switch.
2. Reworked the core selection to logic to first search for the best core and then switch to it if criteria is met.
3. Added logging for relative core errors to XKF5. Created a separate log structure to avoid conflict with EKF2's NKF5.
4. Tested with 3 IMUs in SITL.

Some results from @tridge -

![Screenshot from 2020-06-16 14-41-33](https://user-images.githubusercontent.com/13604433/84756200-1f3f5200-afe0-11ea-99ed-3a1dd23a479b.png)
![Screenshot from 2020-06-16 14-41-55](https://user-images.githubusercontent.com/13604433/84756240-29f9e700-afe0-11ea-99f1-9182e483c32c.png)

Some timing estimates for different thresholds - https://docs.google.com/spreadsheets/d/1UGZopVCcp5afHvm16RUu-mL7smuukvso__V2CIirCl4/edit?usp=sharing

tridge commented on libraries/AP_Logger/LogStructure.h:
```diff
@@ -571,6 +571,24 @@ struct PACKED log_NKF5 {
+    float angErr;
+    float velErr;
+    float posErr;
+    float RErr[3]; // set according to defined maximum allowed EKF3 cores
```
Intead of the new structure I think this should go into XKF3. Sorry for suggesting XKF5 on discord. We can do it as a single float on XKF3 as that msg is per-lane. 
Note that while the XKF3 structure is shared with EKF2, I think that is OK, as we likely will bring across your new lane switching code to EKF2 as well

harshitsankhla replied:
Changed. In the meantime for EKF2, since we are using the same NKF3 structure, setting the error value to be 0. Hope that’s fine for now.

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -7,6 +7,7 @@
 #include <AP_GPS/AP_GPS.h>
 #include <AP_VisualOdom/AP_VisualOdom.h>
 #include <new>
+// #include <stdio.h>
```
should remove this

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -815,34 +822,53 @@ void NavEKF3::UpdateFilter(void)
-        for (uint8_t coreIndex=0; coreIndex<num_cores; coreIndex++) {
+    
+                                   
+    if(runCoreSelection) {
```
should have space after if in ArduPilot style

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -815,34 +822,53 @@ void NavEKF3::UpdateFilter(void)
+                // an alternative core is available for selection based on 2 conditions -
+                // 1. healthy and states have been updated on this time step
+                // 2. has relative error less than primary core error
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
```
I think we should consider dropping the statePredictEnabled check. That check made sense when we were making a decision based on a single sample, but makes less sense with the accumulated error. Paul, what do you think?

priseborough replied:
In practice it should make little difference now given real world noise and sensor errors, but in theory if there is a sensor error transient that effects both lanes, removing it will disadvantage the leading lane. It's an easy change to test so worth a try.

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -907,6 +933,35 @@ void NavEKF3::requestYawReset(void)
+{
+    float error = 0;
+    for (uint8_t i = 0; i < num_cores; i++) {
+        if(i!=primary) {
```
whitespace

tridge commented:
getting close, thanks. We do want one commit per folder, so just run Tools/gittools/git-subsystems-split to split it

tridge commented:
we're also going to need some flight tests of this on a real vehicle. I could fly this on a small quad

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -815,34 +821,53 @@ void NavEKF3::UpdateFilter(void)
+                if(altCoreAvailable) {
+                    // if this core has a significantly lower relative error to the active primary, we consider it as a 
+                    // better core and would like to switch to it even if the current primary is healthy
+                    betterCore = (0.5*CORE_ERR_LIM + altCoreError) < relativeCoreError[primary]; // consider a better core if its relative error is lower by half the max possible error
```
I think the 0.5*CORE_ERR_LIM should be replaced by a separate threshold define. The CORE_ERR_LIM and this threshold are not really related.

tridge commented:
I've been testing this in SITL and I'm not seeing any case where we get the gradual buildup of errors then a switch. I don't know if this is just the thresholds, or if there is something more fundamental.
Note that the EKF logging for plane is only 25Hz. Changing it to be full rate by changing the Log_Write_Fast rate in Plane::scheduler_tasks[] makes it a bit clearer.
This is what I see with a X gyro offset error of 0.02:
![image](https://user-images.githubusercontent.com/831867/85070702-a0bdfc80-b1f9-11ea-85fc-c545273ad9ec.png)
that is with SCHED_LOOP_RATE=400, so each data point is 2.5ms.
the lane switch happened 2s after the parameter change, but we only see the RErr start to build up 12ms before the switch. 
log here:
http://uav.tridgell.net/EKF/00000005.BIN

tridge commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -815,34 +821,53 @@ void NavEKF3::UpdateFilter(void)
+        // 2. is unhealthy
+        // 3. is healthy, but a better core is available
+        // also update the yaw and position reset data to capture changes due to the lane switch
+        if (altCoreAvailable && (primaryErrorScore > 1.0f || !core[primary].healthy() || betterCore)) {
```
we've lost the send_text() for the lane switch which we have in master

harshitsankhla replied:
master doesn't seem to have a send_text() in UpdateFilter(), its in checkLaneSwitch() here - https://github.com/ArduPilot/ardupilot/blob/ca0ae57b56f472ef656ae82eb0a8afe42217c2de/libraries/AP_NavEKF3/AP_NavEKF3.cpp#L899
I'll copy it over. Does this mean that the Lane Switch messages we were seeing before were not because of switches that happened in UpdateFilter() ? I think we might to rework checkLaneSwitch() as well for the new relative error.

tridge commented:
I think it would be helpful to also log errorScore(), so we can see if the relative error is performing as expected

harshitsankhla commented:
Resolved the above !

Some thoughts, should we - 
1. consider having a minimum time required between 2 consecutive switches ?
2. checkLaneSwitch() also be reworked to use the relative error terms ?
3. send_text() if no better lane is found ?

tridge commented:
failing doc tests

harshitsankhla commented:
> failing doc tests

fixed, sorry

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -815,34 +821,54 @@ void NavEKF3::UpdateFilter(void)
+                altCoreAvailable = core[coreIndex].healthy() && statePredictEnabled[coreIndex] && altCoreError < maxCoreError;
+
+                if (altCoreAvailable) {
+                    // if this core has a significantly lower relative error to the active primary, we consider it as a 
```
This will need some follow-on replay tuning to avoid unnecessary resets. Switching does produce some controller transients so we don't want to be doing them frequently.

priseborough commented:
Can we get a branch withe this PR and https://github.com/ArduPilot/ardupilot/pull/14674 so they can be assessed together? I'm not confident I understand how the combination of the two will work.

harshitsankhla commented:
> Can we get a branch with this PR and #14674 so they can be assessed together? I'm not confident I understand how the combination of the two will work.

@priseborough #14674 has this PR's accumulated error logic included. The branch is - https://github.com/harshitsankhla/ardupilot/tree/ekf3-affinity

harshitsankhla commented:
@priseborough 

This is my understanding of the Affinity and Accumulated Error Lane Switching. Also leaving this here for the general audience.

Affinity - For a vehicle which has multiple (Airspeed - Magnetometer - Barometer - GPS) affinity is a way to assign these sensors between separate EKF cores. If a sensor is healthy and can be used for fusion, we want it to be attached to a specific core. If the number of cores is more than the number of sensor instances, then some cores may share the same sensor (specifically the primary one). Considering a vehicle which has 2 baros , 2 GPS, 2 airspeeds, 3 mags and 3 IMUs - following would be the allotment if affinity is enabled and all sensors are healthy.

LANE  |   1  |   2  |  3 
-----------------------------
ASD    |   1  |   2  |   1
BAR    |   1  |   2  |   1
GPS    |   1  |   2  |   1
MAG   |   1  |   2  |   3

Accumulated Error - The error which we accumulate is representative of how bad an EKF core has been over time. So now with affinity, we have a way to determine which sensor combination is the best that can be used. For this, the error term (that we accumulate) should be a right indicator of the sensors performance of that core. 

As Tridge pointed out in his affinity PR for EKF2 #11257  - This provides for a more statistically robust way of using multiple sensors when the airframe is known to have multiple "good" sensors. EKF lane changing is used to select the lane with the best sensors.

@tridge please correct me if I am wrong or have missed something !

tridge commented:
> @tridge please correct me if I am wrong or have missed something !

exactly right, and good explanation. We'll use this in the docs we'll need to write for this feature :-)

tridge commented:
@harshitsankhla as Paul says, we need to setup some tests to see if we have the tuning right. One of the key properties of the lane switching is to save the vehicle on IMU failures. I sent you a log where the first gyro stops responding to vehicle movement on takeoff. The current master lane switching code wasn't fast enough to save the vehicle from flipping over. 
Can you setup some tests in SITL with a few types of major IMU failures. For example:
 - freezing gyro, like in the log I sent you
 - major IMU failure, such as suddenly accel Z axis changes by 5m/s/s or more (which can happen with aliasing)

what we're looking for is to see if the new lane switching scheme is at least as fast as master for these failures.
We also need to organise some real flight tests to ensure we don't get too many lane swicthes

tridge commented:
@harshitsankhla any progress on the testing or test suites? See my comment above from last week

harshitsankhla commented:
> @harshitsankhla any progress on the testing or test suites? See my comment above from last week

here - #14829

harshitsankhla commented:
> @harshitsankhla any progress on the testing or test suites? See my comment above from last week

here - #14829

harshitsankhla commented:
moved to #14674
